azure_iot_edge/datadog_checks/azure_iot_edge/check.py

```diff
@@ -0,0 +1,92 @@
+
+        # Need a custom metric transformer due to version info being located in a JSON-encoded string.
+        edge_agent_metric_transformers = {'edgeAgent_metadata': self._transform_version_metadata}
+        scraper_config = self._edge_agent_check.get_scraper_config(self._config.edge_agent_instance)
+        scraper_config['_default_metric_transformers'].update(edge_agent_metric_transformers)
+
+    def _transform_version_metadata(self, metric, scraper_config):
+        """
+        Submit version metadata from an Edge Agent metadata metric instance.
+
+        Metadata metric look like this:
+
+        ```
+        edgeAgent_metadata{...,edge_agent_version="...",host_information="{\"...\", \"Version\": \"1.0.10~rc2\"}"}
+        ```
+
+        See: https://github.com/Azure/iotedge/blob/1.0.10-rc2/doc/BuiltInMetrics.md#edgeagent
+
+        NOTE: we want the Security Manager version, not the Edge Agent version.
+        """
+        labels = metric.samples[0][OpenMetricsBaseCheck.SAMPLE_LABELS]  # type: dict
+
+        host_information = labels.get('host_information')
+        if host_information is None:
+            self.log.debug('Label "host_information" not found, skipping version metadata')
+            return
+
+        try:
+            host_info = json.loads(host_information)  # type: dict
+        except json.JSONDecodeError as exc:
+            self.log.debug('Error decoding host information: %r', exc)
+            return
+
+        iot_edge_runtime_version = host_info.get('Version')
+        if iot_edge_runtime_version is None:
+            self.log.debug('Key "Version" not found in host_information, skipping version metadata')
+            return
+
+        self.set_metadata('version', iot_edge_runtime_version)
+
+    def check(self, _):
+        self._check_security_manager_health()
+
+        # NOTE: This check consumes configuration from a single instance, and then delegates the Edge Agent and Edge Hub
+        # Prometheus endpoints checks to to two OpenMetrics checks, using a composition approach.
+        # Compared to requiring separate instances for monitor the Edge Agent, Edge Hub, and the Security Manager,
+        # we keep the "1 instance = 1 IoT Edge device" mental model, which provides better configuration UX for users
+        # (as having to configure 3 separate instances is more error-prone).
+        # The composition approach also hopefully makes the code more natural and easier to understand, compared
+        # to subclassing from OpenMetricsBaseCheck and having to dig into its internals such as `.process()` and
+        # scraper configurations.
+
+        # Sync Agent-assigned check ID so that metrics of these sub-checks are reported as coming from this check.
+        self._edge_hub_check.check_id = self.check_id
+        self._edge_agent_check.check_id = self.check_id
+
+        self._edge_hub_check.check(self._config.edge_hub_instance)
```

Comment:
Good catch, not something I've thought about.

> Does run do anything else that we should also do for those two check instances?

`run()` does do some extra stuff, such as honouring options like `set_breakpoint` or `profile_memory` in `init_config`, and returning a `result` string that contains information in case of exception.

https://github.com/DataDog/integrations-core/blob/e9be0e11ee34cc3fdfad665a412dcc3c3b7bac2d/datadog_checks_base/datadog_checks/base/checks/base.py#L797-L839

I don't think there's anything wrong with calling `.run()` on the checks, but we'd need to manage the `result` returned by them (and I'd need to think about how to do that).

Alternatively, I could just abandon the "sub-check" idea and turn `AzureIoTEdgeCheck` into an `OpenMetricsBaseCheck` subclass, with multiple scraper configs. Looks like that might just be the most risk-averse way to go for now.

Reply:
Reorganized the check as an `OpenMetricsBaseCheck`. Turns out to be pretty simple. :+1: